csharp/src/SeedLang.Common/Diagnostic.cs

```diff
@@ -0,0 +1,38 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace SeedLang.Common {
+  // An immutable dignostic info, such as a compiler error or warning.
+  public class Diagnostic {
+    // The name or ID of the reporter.
+    public string Reporter { get; }
```

Comment:
How about define a enum for all the reporters? Otherwise the developer may write wrong reporter string in the code easily.

Reply:
The reporters of a diagnostic could be:

1. System components such as SeedAst Parser, Compiler, Interpreter, etc. These kinds of components can be pre-defined by an Enum type.
1. Third-party extensions. We need to allow third-party extensions to report diagnostics as well. For example, a visualizer extension subscribes to a visualizer provider. When the extension receives a set of data entries from the provider and figures out that there are logical inconsistent issues in the data set, it should have permission to report a diagnostic.

As a reference, VS Code uses a string URI for extensions to report diagnostics.

How about keeping the string type for reporters and providing a set of pre-defined string constants for the system components?